### PR TITLE
Make attribute matching efficient

### PR DESCRIPTION
**Pull request by oehme**
The attribute matching algorithm was using up lots of memory. When doing an up-to-date `assembleDebug` on perf-android-large, it was responsible for 170MB of garbage (out of 2GB). The new algorithm gets along with 10MB instead. 

While documenting the algorithm, a questions came up for me: Do we ever want to allow expensive things like `Map` or `Serializable` as attribute values? If not, we can completely get rid of the attribute value table (which is responsible for the remaining 10MB of memory usage). If we do want to allow it, then it's not a big deal either, there are way bigger fish to fry after this change. The previous algorithm was a bit inconsistent in this regard - It reused the values of the candidates, but re-computed the requested values on every comparison. So it seems we never really made a conscious decision on this before.

### CODE REVIEW

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/attributes/DefaultAttributesSchema.java [hunk 1]
```diff
@@ -157,55 +163,74 @@ public boolean isMatching(AttributeContainerInternal candidate, AttributeContain
...
-        public void disambiguate(Attribute<?> attribute, MultipleCandidatesResult<Object> result) {
+        public Set<Object> disambiguate(Attribute<?> attribute, Object requested, Set<Object> candidates) {
+            DefaultMultipleCandidateResult<Object> result = null;
+
             DisambiguationRule<Object> rules = disambiguationRules(attribute);
-            rules.execute(result);
-            if (result.hasResult()) {
-                return;
+            if (rules.doesSomething()) {
+                result = new DefaultMultipleCandidateResult<Object>(requested, candidates);
```
Comment (bamboo) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/attributes/DefaultAttributesSchema.java]: This exact same block of code appears 4 times, maybe extract?

**Comment by jjohannes** [comment]
@oehme I won't manage to review this today. I'll get to it tomorrow.

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/attributes/DefaultAttributesSchema.java (hunk 1 above)
Comment (oehme) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/attributes/DefaultAttributesSchema.java]: Unfortunately not without reintroducing the garbage creation that this change avoids.

**Comment by oehme** [comment]
Thanks @jjohannes. I'll leave the value table in for now, since in theory we allow expensive values and without the table that would be a performance trap. We can always remove it later.

**Pull request merged**